File: android-native/app/src/main/java/com/example/kpkn/services/workout/PermissionGuideHelper.kt
package com.example.kpkn.services.workout

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.PowerManager
import android.provider.Settings

object PermissionGuideHelper {

    fun isBatteryOptimizationIgnored(context: Context): Boolean {
        val powerManager = context.getSystemService(Context.POWER_SERVICE) as? PowerManager
            ?: return false
        return powerManager.isIgnoringBatteryOptimizations(context.packageName)
    }

    fun isExactAlarmGranted(context: Context): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as? android.app.AlarmManager
            ?: return false
        return alarmManager.canScheduleExactAlarms()
    }

    fun getPermissionState(context: Context): PermissionState {
        return PermissionState(
            batteryOptimizationOk = isBatteryOptimizationIgnored(context),
            exactAlarmOk = isExactAlarmGranted(context),
        )
    }

    fun openBatteryOptimizationSettings(context: Context) {
        try {
            val intent = Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS).apply {
                data = Uri.parse("package:${context.packageName}")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
        } catch (_: Exception) {
            val intent = Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
        }
    }

    fun openExactAlarmSettings(context: Context) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                val intent = Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM).apply {
                    data = Uri.parse("package:${context.packageName}")
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
                context.startActivity(intent)
            }
        } catch (_: Exception) {
            try {
                val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                    data = Uri.parse("package:${context.packageName}")
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
                context.startActivity(intent)
            } catch (_: Exception) { }
        }
    }

    fun openAllPermissionSettings(context: Context) {
        try {
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                data = Uri.parse("package:${context.packageName}")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
        } catch (_: Exception) { }
    }

    data class PermissionState(
        val batteryOptimizationOk: Boolean,
        val exactAlarmOk: Boolean,
    ) {
        val needsAttention: Boolean get() = !batteryOptimizationOk || !exactAlarmOk
        val allOk: Boolean get() = batteryOptimizationOk && exactAlarmOk
    }
}

File: ios-native/KPKNFit/KPKNFit/Services/Workout/PermissionGuideHelper.swift
import Foundation
import UIKit
import UserNotifications

public struct PermissionState {
    public let batteryOptimizationOk: Bool
    public let exactAlarmOk: Bool

    public var needsAttention: Bool { !batteryOptimizationOk || !exactAlarmOk }
    public var allOk: Bool { batteryOptimizationOk && exactAlarmOk }
}

public enum PermissionGuideHelper {

    public static func isBatteryOptimizationIgnored() -> Bool {
        UIApplication.shared.backgroundRefreshStatus == .available
    }

    public static func isExactAlarmGranted() -> Bool {
        true
    }

    public static func getPermissionState() -> PermissionState {
        PermissionState(
            batteryOptimizationOk: isBatteryOptimizationIgnored(),
            exactAlarmOk: isExactAlarmGranted()
        )
    }

    public static func openBatteryOptimizationSettings() {
        guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }

    public static func openExactAlarmSettings() {
        guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }

    public static func openAllPermissionSettings() {
        guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }
}
